kube-runtime/src/scheduler.rs

```diff
@@ -126,9 +138,112 @@ where
     }
 }
 
-/// Stream transformer that schedules each item to be performed at a given time.
+/// Stream transformer that takes a message and `Instant` (in the form of a `ScheduleRequest`), and emits
+/// the message at the specified `Instant`.
+///
+/// Objects are de-duplicated: if a message is submitted twice before being emitted then it will only be
+/// emitted at the earlier of the two `Instant`s.
 pub fn scheduler<T: Eq + Hash + Clone>(
     requests: impl Stream<Item = ScheduleRequest<T>>,
 ) -> impl Stream<Item = Result<T>> {
     Scheduler::new(requests)
 }
+
+#[cfg(test)]
+mod tests {
+    use super::{scheduler, ScheduleRequest};
+    use futures::{channel::mpsc, poll, stream, FutureExt, SinkExt, StreamExt};
+    use tokio::time::{advance, pause, Duration, Instant};
+
+    #[tokio::test]
+    async fn scheduler_should_emit_items_as_requested() {
+        pause();
+        let mut scheduler = scheduler(stream::iter(vec![
+            ScheduleRequest {
+                message: 1u8,
+                run_at: Instant::now() + Duration::from_secs(1),
+            },
+            ScheduleRequest {
+                message: 2,
+                run_at: Instant::now() + Duration::from_secs(3),
+            },
+        ]));
+        assert!(poll!(scheduler.next()).is_pending());
+        advance(Duration::from_secs(2)).await;
+        assert_eq!(scheduler.next().now_or_never().unwrap().unwrap().unwrap(), 1);
```

Comment:
is this a future of result of something? That's a lot of unwrapping

Reply:
1. The internal timer is fallible, so we wrap the `message` in a `Result<T>`
2. `Stream`s can be finite, so `StreamExt::next()` wraps the item as an `Option<Stream::Item>` (returning `None` once the `Stream` is exhausted)
3. `FutureExt::now_or_never()` needs to be able to signal that the future isn't completed yet, so it wraps it as an `Option<Future::Output>` (returning `None` if the `Future` hasn't completed yet)

Normally application code would use `Stream` composition instead, in which case they'd only deal with 1 (the `Result`)